### PR TITLE
[SPARK-19740][MESOS]Add support in Spark to pass arbitrary parameters into docker when running on mesos with docker containerizer

### PR DESCRIPTION
**Pull request by yanji84**
## What changes were proposed in this pull request?

Allow passing in arbitrary parameters into docker when launching spark executors on mesos with docker containerizer @tnachen

## How was this patch tested?

Manually built and tested with passed in parameter

### CODE REVIEW

**Comment by tnachen** [comment]
ok to test

**Comment by tnachen** [comment]
@yanji84 can you add a test for this?

**Comment by yanji84** [comment]
@tnachen I will add test

**Comment by yanji84** [comment]
@tnachen test ready

**Comment by yanji84** [comment]
@tnachen does the PR look good? Thanks!

**Comment by tnachen** [comment]
Hey sorry for the late response, code looks good to me, however we need to add documentation about the new flag. Can you modify the Mesos configuration docs in the docs folder?

**Comment by tnachen** [comment]
@srowen @mgummelt PTAL

File: docs/running-on-mesos.md [hunk 1]
```diff
@@ -357,6 +357,16 @@ See the [configuration page](configuration.html) for information on Spark config
   </td>
 </tr>
 <tr>
+  <td><code>spark.mesos.executor.docker.params</code></td>
+  <td>(none)</td>
+  <td>
+    Set the list of custom parameters which will be passed into the <code>docker run</code> command when launching the Spark executor on mesos using docker containerizer. The format of this property is a comma-separated list of
```
Comment (mgummelt) [inline docs/running-on-mesos.md]: s/mesos/Mesos

s/docker containerizer/the docker containerizer

File: docs/running-on-mesos.md [hunk 2]
```diff
@@ -357,6 +357,16 @@ See the [configuration page](configuration.html) for information on Spark config
...
+    key/value pairs. That is they take the form:
```
Comment (mgummelt) [inline docs/running-on-mesos.md]: s/That is they take the form:/Example:

File: docs/running-on-mesos.md [hunk 3]
```diff
@@ -357,6 +357,16 @@ See the [configuration page](configuration.html) for information on Spark config
   </td>
 </tr>
 <tr>
+  <td><code>spark.mesos.executor.docker.params</code></td>
```
Comment (mgummelt) [inline docs/running-on-mesos.md]: s/params/parameters

to be consistent with the Mesos protobuf terminology

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtil.scala [hunk 4]
```diff
@@ -99,6 +99,26 @@ private[mesos] object MesosSchedulerBackendUtil extends Logging {
     .toList
   }
 
+  /**
+   * Parse a comma-delimited list of arbitrary parameters, each of which
+   * takes the form key=value
+   */
+  def parseParamsSpec(params: String): List[Parameter] = {
+    params.split(",").map(_.split("=")).flatMap { kv: Array[String] =>
```
Comment (mgummelt) [inline resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtil.scala]: s/kv/parameter for naming consistency

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtil.scala [hunk 5]
```diff
@@ -99,6 +99,26 @@ private[mesos] object MesosSchedulerBackendUtil extends Logging {
     .toList
   }
 
+  /**
+   * Parse a comma-delimited list of arbitrary parameters, each of which
+   * takes the form key=value
+   */
+  def parseParamsSpec(params: String): List[Parameter] = {
```
Comment (mgummelt) [inline resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtil.scala]: private

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtil.scala [hunk 6]
```diff
@@ -99,6 +99,26 @@ private[mesos] object MesosSchedulerBackendUtil extends Logging {
     .toList
   }
 
+  /**
+   * Parse a comma-delimited list of arbitrary parameters, each of which
```
Comment (mgummelt) [inline resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtil.scala]: "list of docker parameters"

File: resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtilSuite.scala [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+import org.scalatest._
+import org.scalatest.mock.MockitoSugar
+
+import org.apache.spark.{SparkConf, SparkFunSuite}
+
+class MesosSchedulerBackendUtilSuite extends SparkFunSuite with Matchers with MockitoSugar {
+
+  test("Parse arbitrary parameter to pass into docker containerizer") {
+    val parsed = MesosSchedulerBackendUtil.parseParamsSpec("a=1,b=2,c=3")
+    parsed(0).getKey shouldBe "a"
```
Comment (mgummelt) [inline resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtilSuite.scala]: check out the other test suites.  We use `assert`, not `Matchers`

File: resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtilSuite.scala [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    parsed(0).getValue shouldBe "1"
+    parsed(1).getKey shouldBe "b"
+    parsed(1).getValue shouldBe "2"
+    parsed(2).getKey shouldBe "c"
+    parsed(2).getValue shouldBe "3"
+
+    val invalid = MesosSchedulerBackendUtil.parseParamsSpec("a,b")
```
Comment (mgummelt) [inline resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtilSuite.scala]: invalid entries should be in a separate test

File: resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtilSuite.scala [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+import org.scalatest._
+import org.scalatest.mock.MockitoSugar
+
+import org.apache.spark.{SparkConf, SparkFunSuite}
+
+class MesosSchedulerBackendUtilSuite extends SparkFunSuite with Matchers with MockitoSugar {
+
+  test("Parse arbitrary parameter to pass into docker containerizer") {
+    val parsed = MesosSchedulerBackendUtil.parseParamsSpec("a=1,b=2,c=3")
```
Comment (mgummelt) [inline resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtilSuite.scala]: This should be a private method, so let's not test it.  Just test `containerInfo`

**Comment by yanji84** [comment]
@mgummelt comments addressed, please take another look

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtil.scala [hunk 10]
```diff
@@ -99,6 +99,26 @@ private[mesos] object MesosSchedulerBackendUtil extends Logging {
     .toList
   }
 
+  /**
+   * Parse a list of docker parameters, each of which
+   * takes the form key=value
+   */
+  private def parseParamsSpec(params: String): List[Parameter] = {
+    params.split(",").map(_.split("=")).flatMap { spec: Array[String] =>
```
Comment (mgummelt) [inline resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtil.scala]: s/spec/parameter, for consistency

File: resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtilSuite.scala [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+ */
+
+package org.apache.spark.scheduler.cluster.mesos
+
+import org.scalatest._
+import org.scalatest.mock.MockitoSugar
+
+import org.apache.spark.{SparkConf, SparkFunSuite}
+
+class MesosSchedulerBackendUtilSuite extends SparkFunSuite with Matchers with MockitoSugar {
```
Comment (mgummelt) [inline resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtilSuite.scala]: Is `Matchers` used anymore?

Comment (yanjidriveai) [inline resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtilSuite.scala]: i'll check, if not, will remove it

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtil.scala (hunk 10 above)
Comment (yanjidriveai) [inline resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtil.scala]: to be consistent with other methods, i think we should stick with spec.

>   def parsePortMappingsSpec(portmaps: String): List[DockerInfo.PortMapping] = {
>     portmaps.split(",").map(_.split(":")).flatMap { spec: Array[String] =>
>       val portmap: DockerInfo.PortMapping.Builder = DockerInfo.PortMapping

it seems to be better to reserve the name parameter later for builder

> val param: Parameter.Builder = Parameter.newBuilder()

Comment (mgummelt) [inline resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtil.scala]: ok

**Comment by mgummelt** [comment]
LGTM @srowen ready to merge

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtil.scala (hunk 10 above)
Comment (mridulm) [inline resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtil.scala]: I am not very familiar with docker parameters - but wont this method not fail if '=' exists in the value ?
Or is that prohibited by docker ?

Comment (tnachen) [inline resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtil.scala]: hmm if a value contains a '=' we will have a parsing error.

**Comment by yanji84** [comment]
hello, any update on this?

**Comment by yanji84** [comment]
@srowen is there anything else holding this up? why does it take so long? thanks

**Comment by srowen** [comment]
@yanji84 I have no experience with Mesos and am not, in general, a reviewer for this code, and don't follow Mesos changes. That's why. I pitch in to help merge but would appreciate others stepping up to help.

**Comment by mgummelt** [comment]
@srowen We do appreciate your help with Mesos commits, and generally find you responsive.  I have a habit of pinging you for merges because you seemed to have stepped in once @andrewor14 stepped away.  There are really only two active Spark contributers with knowledge of Mesos code, being myself and @skonto.  We would obviously love it if one of us could be made a committer so we could more effectively maintain the Mesos module.  It's something I've spoked with @rxin about previously, but there was some concern about my total code volume.  I'll be upstreaming a sizable feature this week (Kerberos support on Mesos).  Maybe that, along with my 1.5 year history of responsive code reviews and contributions, would push me over the edge?

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtil.scala (hunk 10 above)
Comment (skonto) [inline resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtil.scala]: It should be quoted: https://github.com/docker/docker/issues/12763

**Comment by tnachen** [comment]
@srowen Appreciate the help you're doing, I think we're doing what we can to help review these patches and making sure Mesos support is still being maintained and improved over time. 
If you trust our judgement and also us still around fixing issues when arises, then we really just need someone like you to help merge patches. 
Ensuring someone else or if anyone that's been contributing to this area can become a committer definitely is a ever ongoing problem that we're still hoping one day can be addressed. Another parallel effort that I think is very worth investigating is to decouple the cluster manager intergation from Spark, which I believe is becoming more relevant now as we have more integration coming.

Long story short, if you can still help in the mean time will be greatly appreciated as we can still make sure improvements around Mesos integration can still happen

**Comment by srowen** [comment]
Is this one still good to go? I see a late comment above. I can merge it, no problem.

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtil.scala (hunk 10 above)
Comment (tnachen) [inline resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtil.scala]: I see, so we should split with a limit instead.
@yanji84 can you fix this?

Comment (yanji84) [inline resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtil.scala]: sorry missed out this comment earlier, Ii set limit to 2, pushed the new pull request

**Comment by yanji84** [comment]
@srowen @tnachen confirm merge

**Comment by srowen** [comment]
Merged to master
